### PR TITLE
Getting commit access

### PR DESCRIPTION
**Pull request by jonmeow**
Establish a process for getting commit access. We will:

-   Grant access based on a developer's commit history.
    -   Someone with commit access should nominate, and a contributor may ask.
    -   A lead will approve nominations. Only one lead is needed.
-   Remove commit access once someone is idle for 6 months.
    -   "Idle" means no significant project activity on any of GitHub, Discord,
        or in meetings.
    -   Access removed due to being idle will be restored on request.

### CODE REVIEW

**Comment by jonmeow** [comment]
Note, looking at [Commit acess](https://github.com/orgs/carbon-language/teams/commit-access):

- camio - last comment Apr 2023, but has been active in other forums
- chandlerc - active this week
- fowles - last comment Oct 2023, no PRs
- geoffromer - active this week
- jonmeow - active this week
- josh11b - active this week
- jsiek - Sep 2022
- mattgodbolt - Jul 2022
- pixep - Aug 2023, though recently active on Discord
- pk19604014 - Sep 2022
- pmqtt - Feb 2023
- tkoeppe - Sep 2021 for PR activity, Nov 2023 in discussions
- wolffg - Aug 2022 for PR activity, active in discussions
- zygoloid - active two weeks ago

So if we're basing this purely on PR activity, that would mean removing commit access for everyone but chandlerc, geoffromer, jonmeow, josh11b, and zygoloid.

File: docs/project/commit_access.md
Comment (chandlerc) [inline docs/project/commit_access.md]: One larger and one smaller thing here...

First, I'd like to emphasize that commit access on Carbon is *not* required for contribution, and really encourage folks to contribute even without it. I've tried to add a paragraph to this effect, but happy for you to rewrite to suit.

A second small thing is that I think reviewing works even w/o commit access, so it's really just approving and merging?

```suggestion
First and foremost, commit access is **not required** for contributing to Carbon! Anyone can send a pull request with a change to Carbon. Very few parts of our process require commit access, and most development activity is exactly the same for folks with or without. This is an important feature of our GitHub workflow for us, and one we plan to keep.

The main thing commit access allows is for developers to merge PRs into the main
[carbon-lang repository](https://github.com/carbon-language/carbon-lang/). They still need review, and often a reviewer with commit access handles the merge. It
also gives the ability to [approve and merge changes](code_review.md).
```

File: docs/project/commit_access.md
Comment (chandlerc) [inline docs/project/commit_access.md]: What do you think about skipping this paragraph? Not sure... But maybe the next paragraph (maybe with my suggestion) can suffice?

File: docs/project/commit_access.md
Comment (chandlerc) [inline docs/project/commit_access.md]: Seems nice to give some guidance on what developers should be thinking about here:

```suggestion
Developers with commit access are expected to make reasonable choices about when
to approve or merge PRs for others -- generally, the change either needs to be trivially understood without context on the code in question (like a cleanup, typo fix, etc), or the developer should have reasonable context on the code in question. If in doubt, feel free to review but leave approving or merging for someone else.

Similarly, developers with commit access are expected to make reasonable choices about what changes to make to their own PRs
without asking for another round of review prior to merge, even after approval. And again, if in doubt, ask for review.
```

File: docs/project/commit_access.md
Comment (chandlerc) [inline docs/project/commit_access.md]: Maybe to just encourage folks to ask a bit more:

```suggestion
After a few non-trivial PRs are merged, contributors should ask a reviewer to nominate for commit access if they plan to keep contributing. Reviewers can also directly suggest and nominate someone.
Nominations need to be approved by at least one lead.
```

File: docs/project/commit_access.md
Comment (chandlerc) [inline docs/project/commit_access.md]:
```suggestion
We'll periodically remove commit access from contributors who have been idle for
over 6 months. We'll use a combination of sources to determine what "idle"
means, including whether a developer has been either sending or reviewing PRs. Developers should always ask for it to be restored if they plan to start contributing again or come back from a break. Any relevant past contributions will still apply and allow it to be restored trivially.

```

**Review by chandlerc — COMMENTED** [review]
Mostly wording tweaks... but maybe @dwblaikie can also help with wording as he's been involved in similar discussions in the LLVM community?

File: docs/project/commit_access.md
Comment (jonmeow) [reply docs/project/commit_access.md]: In the past, every discussion I've had about commit access involved people mentioning they're not fully comfortable doing reviews. That's why it's the one thing I'd been focused on. Do you still want it removed?

(I don't think the next paragraph addresses the question)

File: docs/project/commit_access.md
Comment (jonmeow) [reply docs/project/commit_access.md]: Before:

> Developers with commit access are expected to make reasonable choices about when
> to review and approve PRs for others, and what changes to make to their own PRs
> without asking for another round of review.

After:

> Developers with commit access are expected to make reasonable choices about when
> to approve or merge PRs for others -- generally, the change either needs to be trivially understood without context on the code in question (like a cleanup, typo fix, etc),

Note in this section, I think you changed the focus from "when do you need review for changes to your own PRs" to "when do other PRs need review". I'm trying to adjust phrasing.

Comment (jonmeow) [reply docs/project/commit_access.md]: Nevermind. I think I'm just confused by the emphasis here.

File: docs/project/commit_access.md
Comment (chandlerc) [reply docs/project/commit_access.md]: Interesting! No, I think that's a good reason to keep it. Maybe some slight word smithing?

```suggestion
Note that developers with commit access are not required to do reviews or help with approving and merging PRs. It can still be useful to allow authors to fix trivial PR comments after approval and then merge their own PR, for example because the
review noted a minor typo, or to resolve merge conflicts.

```

File: docs/project/commit_access.md
Comment (zygoloid) [inline docs/project/commit_access.md]: I did not understand this paragraph. (For what it's worth, I think the rest of the text works fine without it.)

First sentence: is this saying
1) Commit access is not accompanied with a responsibility to do reviews or help with approving and merging PRs. (Developers with commit access are not required to [do certain things] _as a condition of their commit access_)
or
2) Reviews and approving or merging PRs can be done without the help of a developer with commit access. (Developers with commit access are not required _in order_ to [do certain things])

Second sentence: What are the consequences of this being useful? Is some particular action encouraged or discouraged, or some policy or advice we want to give here? How is this connected to what we said in the first sentence?

Also, *can* authors without commit access merge their own PRs after approval? I thought that merging PRs, even once approved, still requires commit access?

File: docs/project/commit_access.md
Comment (zygoloid) [inline docs/project/commit_access.md]:
```suggestion
nominate them for commit access if they plan to keep contributing. Reviewers can also
```

File: docs/project/commit_access.md
Comment (zygoloid) [inline docs/project/commit_access.md]:
```suggestion
Developers should always ask for commit access to be restored if they plan to start
```
(The intended antecedent of "it" is two sentences ago, which feels a bit distant.)

Also, is "should always" a bit strong here? Maybe "can" or "should" instead?

File: proposals/p4246.md
Comment (zygoloid) [inline proposals/p4246.md:34]: Is it worth saying somewhere that there may be other reasons why we'd remove commit access, but that they're outside the scope of this proposal? That is -- this proposal isn't trying to say the *only* reason we'd remove commit access is because someone is idle. (It might also be due to some kind of abuse, for example.)

**Review by zygoloid — COMMENTED** [review]
This looks good to me. No real comments on the policy or approach, just some minor wording things.

I do worry a little that having people self-nominate or ask their reviewer for commit access might lead to some awkward conversations if we think they're not ready yet, and might also lead to some forms of self-nomination bias, but I don't have a better suggestion, so I'm happy to go with this.

File: docs/project/commit_access.md
Comment (chandlerc) [reply docs/project/commit_access.md]: FYI, there was some discussion about this paragraph here: https://github.com/carbon-language/carbon-lang/pull/4246#discussion_r1729499088

And...
> Also, _can_ authors without commit access merge their own PRs after approval? I thought that merging PRs, even once approved, still requires commit access?

I believe they cannot -- and I believe enabling this is a primary motivation for granting commit access.

File: proposals/p4246.md
Comment (jonmeow) [reply proposals/p4246.md:34]: Added to background.

Out of curiosity, as lead, were you asking because you were uncertain whether the code of conduct would be superseded this way?

File: docs/project/commit_access.md
Comment (jonmeow) [reply docs/project/commit_access.md]: Done, with "can"

File: docs/project/commit_access.md
Comment (jonmeow) [reply docs/project/commit_access.md]: Done; note, also "can" on the line above.

File: docs/project/commit_access.md
Comment (jonmeow) [reply docs/project/commit_access.md]: Authors without commit access haven't ever been able to merge changes.

Also, recently actions became more restrictive for security reasons. I think in the past, a contributor only needed "approve and run" on their first PR. I'm not sure if there's _any_ threshold where "approve and run" is no longer needed now.

So the present state is that people without commit access require assistance in order to _both_ run actions _and_ merge commits.

Comment (jonmeow) [reply docs/project/commit_access.md]: Note, rephrased the paragraph again.

**Review by zygoloid — COMMENTED** [review]
LGTM

File: proposals/p4246.md
Comment (zygoloid) [reply proposals/p4246.md:34]: Thanks!

You've addressed this to my satisfaction, but since you asked, here are a couple of examples of situations that I'd like to avoid:

- Someone behaves badly on github, let's say by merging PRs that it's clear should not have been merged, and we choose to remove their commit access, and they point at this proposal and say "you're not following your own documented rules" / "you're mistreating me".
- Someone looks at this proposal and decides they don't want to participate in our community because they see no remedy for people who misuse their commit access.

I think just something minimal -- like you've added -- that clarifies that these are not the *only* rules governing who gets to have commit access is sufficient.

File: docs/project/commit_access.md
Comment (zygoloid) [reply docs/project/commit_access.md]: Thank you! The new phrasing works for me.

Comment (jonmeow) [reply docs/project/commit_access.md]: Sorry, one small edit, "smooth over" -> "ease" due to possible connotations.

**Review by chandlerc — APPROVED** [review]
Approving for leads (and confirmed with all three FWIW, given the nature of this). Thanks!

**Pull request merged**
